kong/core/router.lua

```diff
@@ -128,18 +143,48 @@ local function marshall_api(api)
     if #api.uris > 0 then
       api_t.match_rules = bor(api_t.match_rules, MATCH_RULES.URI)
 
-      for i, uri in ipairs(api.uris) do
-        local escaped_uri = [[\Q]] .. uri .. [[\E]]
-        local strip_regex = escaped_uri .. [[\/?(.*)]]
+      for _, uri in ipairs(api.uris) do
+        if re_match(uri, [[^[a-zA-Z0-9\.\-_~/%]*$]]) then
+          -- plain URI or URI prefix
+          local escaped_uri = [[\Q]] .. uri .. [[\E]]
+          local strip_regex = escaped_uri .. [[/?(?P<stripped_uri>.*)]]
+
+          api_t.uris[uri] = {
+            prefix        = true,
+            strip_regex   = strip_regex,
+          }
+
+          insert(api_t.uris, {
+            prefix      = true,
+            value       = uri,
+            regex       = escaped_uri,
+            strip_regex = strip_regex,
+          })
+
+        else
+          -- regex URI
+          local strip_regex = uri .. [[/?(?P<stripped_uri>.*)]]
+
+          local from, _, err = re_find(uri, [[\(.*\)]], "jo")
```

Comment:
We should even consider using Lua patterns for this instead, as the PCRE is overkill for such a use-cases.